### PR TITLE
Add usage limits fields for new discount system

### LINKED ISSUE #38657: Set a usage limit per customer
# Context
The user can set a usage limit per customer (integer >= 1).
By default, this limit is 1.
An option allows for unlimited usage per customer.
If a discount is intended to a single customer, this option is not displayed.
Once the limit of use/application is reached (https://github.com/PrestaShop/PrestaShop/issues/38663), customers can no longer benefit from the discount, even if the limit of usage for those customers is not reached.

# Mockup

Image

# Tests
## Field validation
**GIVEN:** an invalid usage limit 
**WHEN:** I want to create the discount
**THEN:** the discount is not created and a error message is displayed

**GIVEN:** a valid usage limit
**WHEN:** I want to create the discount
**THEN:** the discount is created

**GIVEN:** a limit of use set to 50
**WHEN:** I set a usage limit to 49
**THEN:** the customer limit is valid

**GIVEN:** a limit of use set to 50
**WHEN:** I set a usage limit to 51
**THEN:** the customer limit is not valid

**GIVEN:** a discount intended to a single customer
**WHEN:** I want to set a usage limit
**THEN:** the field is not diplayed

## Apply discount
**GIVEN:** the usage limit for a customer is reached
**WHEN:** I want to apply the discount via promo code (with the same customer)
**THEN:** the discount is not applied and a error message is displayed instead

**GIVEN:** a discount automatically applied
**WHEN:** the usage limit for a customer is reached 
**THEN:** the discount is not applied

**GIVEN:** the usage limit for a customer is not reached yet
**WHEN:** I want to apply the discount via promo code (with the same customer)
**THEN:** the discount is applied

**GIVEN:** a discount applied automatically
**WHEN:** the usage limit for a customer is not reached yet
**THEN:** the discount is applied

**GIVEN:** the usage limit for a customer is not reached yet, but the limit of use/application is reached
**WHEN:** I want to apply the discount via promo code (with the same customer)
**THEN:** the discount is not applied and a error message is displayed instead

**GIVEN:** a discount automatically applied
**WHEN:** the usage limit for a customer is not reached yet, but the limit of use/application is reached
**THEN:** the discount is not applied

### LINKED ISSUE #38663: Set a limit on the application/use of discounts
# Context
Merchants can specify whether a discount can be applied/used without limitation or set a limit.
This limit must be an integer greater than or equal to 1 (1 by default).

If the discount is intended for all customer or specific customer group(s), only the first X (= limit) customers could benefit from the discount.
If the discount is intended for a single customer, only the same customer could benefit X (=limit) time(s) from the discount.

So here we in the end have 3 cases :

    limit total per discount => 100 uses that can all be used by any user ( one or more )
    limit per user => no uses limit on the discount but each user can use it x amount of time
    the combination of both => common pile => I have 100 discount available each user can use it 20 times => if there is only 15 discount left the last user can only use it 15 times.

Design wise, both field should be placed next to each other and the wording of the per customer ( 1st case ) should be something like "total uses of discounts"

# Mockup
TBUP

# Tests
## Field validation
**GIVEN:** an invalid limit 
**WHEN:** I want to create the discount
**THEN:** the discount is not created and a error message is displayed

**GIVEN:** a valid limit
**WHEN:** I want to create the discount
**THEN:** the discount is created

## Apply discount
**GIVEN:** the limit is reached
**WHEN:** I want to apply the discount via promo code
**THEN:** the discount is not applied and a error message is displayed instead

**GIVEN:** a discount automatically applied
**WHEN:** the limit is reached 
**THEN:** the discount is not applied

**GIVEN:** the limit is not reached yet
**WHEN:** I want to apply the discount via promo code
**THEN:** the discount is applied

**GIVEN:** a discount applied automatically
**WHEN:** the limit is not reached yet
**THEN:** the discount is applied

### PR DESCRIPTION
**Pull request by boherm**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.1.x
| Description?      | Add usage limits fields for new discount system.<br><br>**Bonus:** add the ability to add a custom label for disabling_switch component.<br/>**Bonus2:** introduce a DirtyTrait to track modified properties in an object.<br>**Bonus3:** add tests for usage limits in FO (and validate the `null` state too for no limit!)
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | See #38663 & #38657
| UI Tests          | https://github.com/boherm/ga.tests.ui.pr/actions/runs/20273313372
| Fixed issue or discussion?     | Fixes #38663 Fixes #38657
| Related PRs       | https://github.com/PrestaShop/ps_apiresources/pull/119
| Sponsor company   | PrestaShop SA

### CODE REVIEW

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `branch` should be `develop`, `9.0.x` or `8.2.x`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!

(Note: this is an automated message, but answering it will reach a real human)

**Comment by ps-jarvis** [comment]
This pull request seems to contain new translation strings. I have summarized them below to ease up review:

- `Admin.Catalog.Feature`
    - [ ] `Specifiy discount mode`
    - [ ] `Select total usage limits`
    - [ ] `No limit`
    - [ ] `Select usage limits per customer`
- `Admin.Catalog.Help`
    - [ ] `This discount can be used "X" time(s) in total.`
    - [ ] `A customer will only be able to use this discount "X" time(s).`
- `Admin.Catalog.Notification`
    - [ ] `The usage limit per customer cannot be higher than the total usage limit.`

(Note: this is an automated message, but answering it will reach a real human)

**Review by jolelievre — DISMISSED** [review]

**Review by tleon — DISMISSED** [review]

**Review by nicosomb — DISMISSED** [review]

**Review by paulnoelcholot — COMMENTED** [review]
Hello @boherm 

I tested your PR and after enabled discount V2, when I go to the Discount page I got that error : 
image

Can you check that? :/

**Comment by boherm** [comment]
> Can you check that? :/

Hi @paulnoelcholot, it's not related to this PR. You need to reinstall your database 😄

**Comment by boherm** [comment]
PR for autoupgrade: https://github.com/PrestaShop/autoupgrade/pull/1581

**Review by jolelievre — DISMISSED** [review]

**Review by mattgoud — DISMISSED** [review]

**Review by SiraDIOP — COMMENTED** [review]
Hello @boherm,

Thanks for your PR, i tested it and i have 2 issues:
When I set the limit to 1 and place an order, the discount is applied, but I can reuse the code even though usage was at 1. The bug affects free shipping and catalog products.

https://github.com/user-attachments/assets/d6738bc9-f7fb-45f0-ae0b-9ceabf60d426


https://github.com/user-attachments/assets/57a4bd9d-44fc-49b9-bcf6-2325aa3c0ebd

**Review by jolelievre — APPROVED** [review]

**Review by nicosomb — APPROVED** [review]

**Comment by boherm** [comment]
Hi @SiraDIOP!

I've got some news about your 2 problems :)

After some tests, for the product level discount : 
- First, a product_level discount must have a product segment or single product defined.
- Then, if you apply a product_level discount in your cart that haven't the right product in it, this discount are in the cart but aren't used
- And so, the total quantity doesn't decrease and we don't considering this discount to be used.

For free shipping discount : 
- If you select a carrier that is already free before apply a free_shipping discount
- The total quantity doesn't decrease and we don't considering this discount to be used.

So, in fact, what you describe, looks legit to me in fact! 😄

(I had some tests with product_level and free_shipping to validate the main feature of theses discounts in FO)

**Comment by SiraDIOP** [comment]
Hello @boherm,

Thanks for your PR, i tested it and it's good for me ^^
for issue https://github.com/PrestaShop/PrestaShop/issues/38657 
GIVEN: an invalid usage limit
WHEN: I want to create the discount
THEN: the discount is not created and a error message is displayed ✅

GIVEN: a valid usage limit
WHEN: I want to create the discount
THEN: the discount is created ✅

GIVEN: a limit of use set to 50
WHEN: I set a usage limit to 49
THEN: the customer limit is valid ✅

GIVEN: a limit of use set to 50
WHEN: I set a usage limit to 51
THEN: the customer limit is not valid ✅

GIVEN: a discount intended to a single customer
WHEN: I want to set a usage limit
THEN: the field is not diplayed ✅

GIVEN: the usage limit for a customer is reached
WHEN: I want to apply the discount via promo code (with the same customer) 
THEN: the discount is not applied and a error message is displayed instead ✅

GIVEN: a discount automatically applied
WHEN: the usage limit for a customer is reached
THEN: the discount is not applied ✅

GIVEN: the usage limit for a customer is not reached yet
WHEN: I want to apply the discount via promo code (with the same customer)
THEN: the discount is applied ✅

GIVEN: a discount applied automatically
WHEN: the usage limit for a customer is not reached yet
THEN: the discount is applied ✅

GIVEN: the usage limit for a customer is not reached yet, but the limit of use/application is reached
WHEN: I want to apply the discount via promo code (with the same customer)
THEN: the discount is not applied and a error message is displayed instead ✅

GIVEN: a discount automatically applied
WHEN: the usage limit for a customer is not reached yet, but the limit of use/application is reached
THEN: the discount is not applied ✅

For https://github.com/PrestaShop/PrestaShop/issues/38663
Field validation
GIVEN: an invalid limit
WHEN: I want to create the discount
THEN: the discount is not created and a error message is displayed ✅

GIVEN: a valid limit
WHEN: I want to create the discount
THEN: the discount is created ✅

Apply discount
GIVEN: the limit is reached
WHEN: I want to apply the discount via promo code
THEN: the discount is not applied and a error message is displayed instead ✅

GIVEN: a discount automatically applied
WHEN: the limit is reached
THEN: the discount is not applied ✅

GIVEN: the limit is not reached yet
WHEN: I want to apply the discount via promo code
THEN: the discount is applied ✅

GIVEN: a discount applied automatically
WHEN: the limit is not reached yet
THEN: the discount is applied ✅

**Review by SiraDIOP — APPROVED** [review]

File: classes/CartRule.php [hunk 1]
```diff
@@ -131,8 +131,8 @@ class CartRuleCore extends ObjectModel
             'date_from' => ['type' => self::TYPE_DATE, 'validate' => 'isDate', 'required' => true],
             'date_to' => ['type' => self::TYPE_DATE, 'validate' => 'isDate', 'required' => true],
             'description' => ['type' => self::TYPE_STRING, 'validate' => 'isCleanHtml', 'size' => DiscountSettings::MAX_DESCRIPTION_LENGTH],
-            'quantity' => ['type' => self::TYPE_INT, 'validate' => 'isUnsignedInt'],
-            'quantity_per_user' => ['type' => self::TYPE_INT, 'validate' => 'isUnsignedInt'],
+            'quantity' => ['type' => self::TYPE_INT, 'allow_null' => true, 'validate' => 'isUnsignedInt'],
```
Comment (Hlavtox) [inline classes/CartRule.php:134]: BC break

File: classes/CartRule.php [hunk 2]
```diff
@@ -450,7 +450,7 @@ public static function getCustomerCartRules(
         // Then, conditions for date, voucher active property and total amount of vouchers in stock
         $sql .= ' AND NOW() BETWEEN cr.date_from AND cr.date_to
             ' . ($active ? 'AND cr.`active` = 1' : '') . '
-            ' . ($inStock ? 'AND cr.`quantity` > 0' : '');
+            ' . ($inStock ? 'AND (cr.`quantity` > 0 OR cr.`quantity` is null)' : '');
```
Comment (Hlavtox) [inline classes/CartRule.php:453]: That's why you had to do this

File: classes/CartRule.php [hunk 3]
```diff
@@ -797,7 +797,7 @@ public function checkValidity(Context $context, $alreadyInCart = false, $display
             }
 
             // We verify the total available quantity
-            if (!$this->quantity) {
+            if ($this->quantity !== null && !$this->quantity) {
```
Comment (Hlavtox) [inline classes/CartRule.php:800]: and this

File: classes/CartRule.php [hunk 4]
```diff
@@ -840,7 +840,7 @@ public function checkValidity(Context $context, $alreadyInCart = false, $display
                     ++$quantityUsed;
                 }
 
-                if ($quantityUsed > $this->quantity_per_user) {
+                if ($this->quantity_per_user !== null && $quantityUsed > $this->quantity_per_user) {
```
Comment (Hlavtox) [inline classes/CartRule.php:843]: and this

File: classes/CartRule.php [hunk 5]
```diff
@@ -1991,7 +1991,7 @@ public static function autoAddToCart(?Context $context = null, bool $useOrderPri
 		LEFT JOIN ' . _DB_PREFIX_ . 'cart_rule_country crco ON cr.id_cart_rule = crco.id_cart_rule
 		WHERE cr.active = 1
 		AND cr.code = ""
-		AND cr.quantity > 0
+		AND (cr.quantity > 0 OR cr.quantity is null)
```
Comment (Hlavtox) [inline classes/CartRule.php:1994]: and this

File: classes/PaymentModule.php [hunk 6]
```diff
@@ -1324,7 +1324,10 @@ protected function createOrderCartRules(
 
                 // Create a new instance of Cart Rule without id_lang, in order to update its quantity
                 $cart_rule_to_update = new CartRule((int) $cartRule->id);
-                $cart_rule_to_update->quantity = max(0, $cart_rule_to_update->quantity - 1);
+                // if the cart rule isn't in nolimit mode, we decrease the quantity available
+                if ($cart_rule_to_update->quantity !== null) {
```
Comment (Hlavtox) [inline classes/PaymentModule.php:1328]: and this

File: classes/CartRule.php (hunk 1 above)
Comment (jolelievre) [reply classes/CartRule.php:134]: Allowing null is not a BC break, previous values are still retro compatible Import/export will work as it always did It's just that the column accepts a new value `null` that has a new meaning `unlimited`

So I disagree with you on this one

File: classes/CartRule.php (hunk 2 above)
Comment (jolelievre) [reply classes/CartRule.php:453]: I also disagree on this

File: classes/CartRule.php (hunk 3 above)
Comment (jolelievre) [reply classes/CartRule.php:800]: and this

File: classes/CartRule.php (hunk 4 above)
Comment (jolelievre) [reply classes/CartRule.php:843]: and this

File: classes/CartRule.php (hunk 5 above)
Comment (jolelievre) [reply classes/CartRule.php:1994]: and this

File: classes/PaymentModule.php (hunk 6 above)
Comment (jolelievre) [reply classes/PaymentModule.php:1328]: and this :trollface:

File: classes/CartRule.php (hunk 1 above)
Comment (Hlavtox) [reply classes/CartRule.php:134]: @jolelievre Jo, imagine this use case.

You have a module that does something, expects cart rules to have an integer. Now, since 9.1, there can be nulls suddenly. Your module stops working.

That's a BC break no?

**Comment by Touxten** [comment]
If you read the ADR, there doesn't seem to be a BC Break.

However, it hasn't been updated in a long time.

https://github.com/PrestaShop/ADR/blob/master/0017-backward-compatibility-promise.md#database-structure

We should consult again to clarify this.

Basically, I agree with @Hlavtox  that a null value could cause errors on third-party modules. In any case, this requires documentation.

**Comment by Touxten** [comment]
We could also discuss this point. 

https://github.com/PrestaShop/ADR/blob/master/0017-backward-compatibility-promise.md#php-methods-and-functions

File: classes/CartRule.php (hunk 1 above)
Comment (kpodemski) [reply classes/CartRule.php:134]: @Hlavtox If you're doing things right, you type it to `int` anyway, we can document this, but let's not go to extremes, we care A LOT about backward compatible things, but this cannot overshadow our need for development, and if there's little subset of users that will have to adjust the code between one minor and another, let's be it

**Comment by kpodemski** [comment]
@Touxten 

I added a proper label to document this, at the moment, I don't see a need to discuss this ADR again.

**Comment by Touxten** [comment]
Cool ❤️ 

` I don't see a need to discuss this ADR again.`

Because return type is changed and so it's a BC Break if we read the ADR

Screenshot - 2026-01-16T151234 256

For the change in DB it's ok but not for the return change.

I think it could be constructive.

PrestaShop developers, module developers, e-merchants—we don't all have the same vision of BC Break.

**Comment by Touxten** [comment]
I agree with you about the improvements; we mustn't block everything, which is why I propose a constructive discussion

**Comment by kpodemski** [comment]
@Touxten 

what to me is important, is that we should try to be more pragmatic

of course, there is a way to make it BC Break-free, but that would:
- be much more complex task
- more time-consuming

unless @Hlavtox have thought about it when he brought this _issue_

there's one more thing, even tho the quantity may be `null`, such code:

```
<?php

$quantity = null;

if ($quantity < 1) { echo 'test'; }
```

would still work, it will produce a deprecation warning on PHP 8.1+, but that's all

**Comment by Touxten** [comment]
On a module: 

```
if (is_int($quantity)) {
   ... ...
}
```

Marking a PR as requiring documentation or a compatibility break is not a failure, but rather an improvement. "We couldn't find the resource,  we can't make it any simpler at the moment, that's why... "

It also proves that people are taking the time to examine the code and see what could block modules for future versions, which is a guarantee of quality for me.

As you said above, we can't block everything to make improvements.

PS : the code is stupid because is a int but i see that on a lot of module
